### ISSUE TITLE
rutorrent http error 0

### ISSUE THREAD

**Issue by Redtwist27** [bogenpirat/remote-torrent-adder#260]
Hello

been trying to find a solution to remote torrent adder stating http error 0 when instead torrent are correctly added to rutorrent

any idea?

**Comment by bogenpirat** [comment]
check `chrome://extensions/?id=oabphaconndgibllomdcjbfdghcmenci`, go to "inspect views" and click on "background page". any errors there when you get the "http error 0" message?

**Comment by Redtwist27** [comment]
i see this:

Access to XMLHttpRequest at 'http://XXXXX:XXXXXX@192.XXX.XX.XXX:XXXX/php/addtorrent.php?' from origin 'chrome-extension://oabphaconndgibllomdcjbfdghcmenci' has been blocked by CORS policy: Redirect location '' contains a username and password, which is disallowed for cross-origin requests.
ruTorrentWebUI.js:45 POST http://username :password?@192.XXX.XX.XXX:XXXX/php/addtorrent.php? net::ERR_FAILED
RTA.clients.ruTorrentAdder @ ruTorrentWebUI.js:45
RTA.dispatchTorrent @ functions.js:24
RTA.getTorrent @ functions.js:56
(anonymous) @ background.js:76

**Comment by bogenpirat** [comment]
odd. any chance you have an old chrome version? we recently had issues wrt CORS restrictions. i've got my rutorrent setup running fine with chrome 76.0.3809.132.

**Comment by Redtwist27** [comment]
running on brave Version 0.68.140 Chromium: 77.0.3865.90

**Comment by bogenpirat** [comment]
`net::ERR_FAILED` leads me to believe that this stems from a network connectivity issue. in that case, the torrents should not be added, though... does the error occur when you disable your ad blocker (or similar extensions that tamper with the webrequest api)?

**Comment by Redtwist27** [comment]
tested with all the ad block function disabled via the setting tab problem still persist. will try with chrome soon

**Comment by meanddem** [comment]
I have this same error and I am using an updated version of google chrome. Any new ideas how to fix this?

**Comment by Redtwist27** [comment]
i guess no use to test myself then? xD

**Comment by bogenpirat** [comment]
give it a try anyway. as it stands, i can't reproduce this, which is developer-speak for "i don't see this happening".

if it doesn't happen anymore, then the two of you have made the same configuration error (likely on the RTA side). if it still happens, it might be related to your rutorrent setup. in both cases i can try and see if i can work around it in the code.

in either case, however, i can't fix what i can't see.

**Comment by mudboy** [comment]
I am seeing the same error, I use Brave but installed the latest Chrome (77.0.3865.90) to test that and it fails with the same error.

`Access to XMLHttpRequest at '<target-uri>' from origin 'chrome-extension://oabphaconndgibllomdcjbfdghcmenci' has been blocked by CORS policy: Redirect location '' contains a username and password, which is disallowed for cross-origin requests.`

Something about CORS has been changed in the latest chromium builds. It doesn't like usernames and passwords in cross-origin requests. I was trying to via rutorrent which sends the username and password in the request uri.

**Comment by bogenpirat** [comment]
i'm using chrome, too. Version 77.0.3865.90 (Official Build) (64-bit). i'm not getting this error.

the thing is, there was a bit of a discussion going in #258 and the last comments made it seem like the chrome team reverted the change that led to this.

there's some discussion in #198 about embedded credentials in xmlhttprequests (nothing to do with CORS per se) that from skimming over it reads like they saw reason and reverted back to the original functionality, but that's from 2017 and targeting chrome 59.

there's a more recent change to CORS policy in chrome extensions where a **content script** is no longer allowed to make cross origin requests, but this doesn't apply to RTA because RTA does the cross origin request in its background page, which is explicitly still functional as per the developer note.

so i'm still at a loss here.

**Comment by Redtwist27** [comment]
Version 77.0.3865.90 (Official Build) (64-bit)

same error

**Comment by bogenpirat** [comment]
alright, here's an idea: go to your rutorrent webui directly, hit f12, go to the network tab, then *manually* add any .torrent file to it that you have on your computer, e.g. from [here](http://releases.ubuntu.com/19.04/ubuntu-19.04-desktop-amd64.iso.torrent).

this will create an entry in the devtools to addtorrent.php with method POST and status code 302, and **should** add another to addtorrent.php with a bunch of parameters added afterwards with method GET and a status code of 200.

can either of you a) confirm that this is the case, and b) get me the list of headers you get when clicking on ideally both, but definitely the latter of these two?

and just to confirm: for all three of you, the behavior is that the torrent is added correctly, but you get the error message that it wasn't?

**Comment by Redtwist27** [comment]
i'm adding magnet not torrent file.

**Comment by bogenpirat** [comment]
then do that instead of adding a file.

**Comment by Redtwist27** [comment]
i see this:

first one:

HTTP/1.1 302 Moved Temporarily
Server: nginx/1.16.1
Date: Sun, 29 Sep 2019 13:06:10 GMT
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive
X-Powered-By: PHP/7.3.9
Location: addtorrent.php?result[]=Success&name[]=


then:

Request URL: http://XXXXXXXX:XXXXX/php/addtorrent.php?result[]=Success&name[]=magnetxxxxxxxxxxxx
Request Method: GET
Status Code: 200 OK
Remote Address: XXXXX
Referrer Policy: no-referrer-when-downgrade

**Comment by bogenpirat** [comment]
okay, then i'll need to know where exactly the error message comes from. please go to `chrome://extensions/?id=oabphaconndgibllomdcjbfdghcmenci`, click on background page under inspect views and the network tab. then try adding a torrent (which should result in an error). let me know whether there's an error in the network tab (besides the one displayed in the console below).

**Comment by gcaliber** [comment]
This was happening to me, but it works fine again with newer versions of chrome.

**Comment by madmaxm** [comment]
Hi,  
I have the HTTP error 0 pop up as well when trying to add a .torrent via link catching.

<img width="281" alt="RTA-Irregular_HTTP_Code_0" src="https://user-images.githubusercontent.com/128876/68659747-21b46780-0538-11ea-9bb6-5060b12a0368.PNG">

Looking for errors on the background page, I can see the following:
ruTorrentWebUI.js:65 POST https://xxx.xxx/rutorrent/php/addtorrent.php?label=APPS 401 (Unauthorized)
XMLHttpRequest.sendAsBinary @ functions.js:9
RTA.clients.ruTorrentAdder @ ruTorrentWebUI.js:65
RTA.dispatchTorrent @ functions.js:24
xhr.onreadystatechange @ functions.js:69
XMLHttpRequest.send (async)
RTA.getTorrent @ functions.js:76
(anonymous) @ background.js:76
_generated_background_page.html:1 Access to XMLHttpRequest at 'https://user:password@xxx.xxx/rutorrent/php/addtorrent.php?label=APPS' from origin 'chrome-extension://oabphaconndgibllomdcjbfdghcmenci' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

What can I do?

**Comment by sieempi** [comment]
This happens because the extension is not sending basic authentication header, which is required by some rutorrent installations.
The sollution is adding the header in ruTorrentWebUI.js:27
xhr.open("POST", url, true, server.login, server.password);
->xhr.setRequestHeader('Authorization','Basic ' + Base64StringOfUserColonPassword);

**Comment by ohhdemgirls** [comment]
> This happens because the extension is not sending basic authentication header, which is required by some rutorrent installations.
> The sollution is adding the header in ruTorrentWebUI.js:27
> xhr.open("POST", url, true, server.login, server.password);
> ->xhr.setRequestHeader('Authorization','Basic ' + btoa(server.login + ":" + server.password));

Where, how, what? I've looked at the `ruTorrentWebUI.js` and it's in there but [not on line 27](https://github.com/bogenpirat/remote-torrent-adder/blob/master/webuiapis/ruTorrentWebUI.js#L73). It's extremely fucking annoying that there's a clear fix for this but it's not being implemented. And it looks like things have changed again because the torrent doesn't add to my client despite the error like some are reported. When you manage 40+ seedboxes and lose this functionality..... I'm tearing my hair out.

How about you make clear how to implement this beginning to end rather than ref code that kills the conversation because.... ohh ffs, w/e ballocks.

**Comment by gcaliber** [comment]
> This happens because the extension is not sending basic authentication header, which is required by some rutorrent installations.
> The sollution is adding the header in ruTorrentWebUI.js:27
> xhr.open("POST", url, true, server.login, server.password);
> ->xhr.setRequestHeader('Authorization','Basic ' + btoa(server.login + ":" + server.password));

I changed this locally and it made no difference for me. 

I found that setting chrome://flags/#out-of-blink-cors to disabled fixes it for me.

**Comment by takaacstamaas** [comment]
Alternative solution:

**Xirvik**
https://chrome.google.com/webstore/detail/xirvik-torrent-to-seedbox/gljdkkichjgocpdmiaachhlfccddcjgb

It works with rtorrent as well.

**Comment by Nico83500** [comment]
Hi, we are several with this issue, it returns 0 but it adds the torrent to rutorrent. If I disable Out of blink CORS, it works. I just reinstall my computer and the issue is the same.

**Comment by snvoid** [comment]
I'm experiencing the same error while trying to use rutorrent with my seedbox provider. The torrent does get added even though it shows an error instead of the green success. It was working fine 2 days ago, so nothing really has changed on my end. 

Chrome - Version 79.0.3945.130 (Official Build) (64-bit)

**Comment by winedog** [comment]
Can confirm this issue is happening again for me on Version 80.0.3987.87 (Official Build) (64-bit)

**Comment by dabl2** [comment]
> Can confirm this issue is happening again for me on Version 80.0.3987.87 (Official Build) (64-bit)

Same here.

**Comment by meanddem** [comment]
Likewise, same issue as well.

**Comment by Litz97** [comment]
Same here - Works ok with deluge installed on same server. Rutorrent seems to be the problem here.

**Comment by snvoid** [comment]
Seems to have stopped working entirely now. Seems like the latest version of Chrome completely made this extension non functional. Unfortunately the other extension Xirvik has stopped working as well.

**Comment by winedog** [comment]
> Seems to have stopped working entirely now. Seems like the latest version of Chrome completely made this extension non functional. Unfortunately the other extension Xirvik has stopped working as well.

Still working for me, but it is throwing the error too.  I'm just ignoring the error at this time.  I have noticed that everyone once in awhile it won't add the file, but I can't replicate it consistently.

What Chrome version are you on?  I'm on  80.0.3987.106 (Official Build) (64-bit)

**Comment by WolvesofWar** [comment]
> I found that setting chrome://flags/#out-of-blink-cors to disabled fixes it for me.

Second this fix. Rutorrent works with extension now.

**Comment by snvoid** [comment]
> > I found that setting chrome://flags/#out-of-blink-cors to disabled fixes it for me.
> 
> Second this fix. Rutorrent works with extension now.

This worked for me as well! Hopefully there can be a more permanent fix for this.

**Comment by nero01** [comment]
Hi, i found a working fix for the CORS policy issue that i got with the latest update.
The extension has to be extracted with a tool such as CRX viewer and the following file has to be edited: ruTorrentWebUI.js
Look for the line "xhr.open("POST", url, true, server.login, server.password);" and replace it with 
`xhr.open("POST", url, true, server.login, server.password);
  xhr.setRequestHeader("Access-Control-Allow-Origin", "*");`

**Comment by winedog** [comment]
> Hi, i found a working fix for the CORS policy issue that i got with the latest update.
> The extension has to be extracted with a tool such as CRX viewer and the following file has to be edited: ruTorrentWebUI.js
> Look for the line "xhr.open("POST", url, true, server.login, server.password);" and replace it with
> `xhr.open("POST", url, true, server.login, server.password); xhr.setRequestHeader("Access-Control-Allow-Origin", "*");`
> 
> This code change will allow all origins for the CORS policy
> 
> Love to see that the addon shows success when a torrent is added :)

Tried this and it didn't seem to do anything for me? I'm on Version 80.0.3987.163 (Official Build) (64-bit)

**Comment by kionez** [comment]
@JPRuskin solution works fine here (rutorrent + nginx basic auth + chrome 81.0.4044.122)

**Comment by rmuchall** [comment]
As there was some delay in resolving this issue I had a go at coding a similar extension in TypeScript.
https://chrome.google.com/webstore/detail/torrent2box/cpgkjlmailbidmlkafnfjehoimkkpabn
(only works for ruTorrent right now)

**Comment by bogenpirat** [comment]
if everyone's issue was indeed just the authentication (which the error message would **not** indicate), then that's fixed in repo code now.

i wrote some code that may fix CSRF-related issues for *some* ruTorrent installations here, give that a try: https://github.com/bogenpirat/remote-torrent-adder/issues/262#issuecomment-667723767

this issue has become a bit cluttered with people posting who have likely separate issues. please feel free to open a new issue if this problem persists in the new release that will soon be put out.
